### ISSUE TITLE
[WIP] Automation design

### ISSUE THREAD

**Issue by sharwell** [dotnet/xliff-tasks#42]
🚧 This issue is a work-in-progress design document for improving automation related to the XLF files in a repository.

### Developer goals

* Contributors must not be required to create, edit, commit, merge, or review XLF files unless they are contributing strings in a non-English language.
* Changes to RESX files (and other original string sources) must not result in local or CI build failures prior to a localization freeze of the target branch.
* Contributors should be able to submit updated translation(s) for a localization unit by modifying the relevant localization units in the XLF file(s) of interest, without incurring any other changes

### Localization team goals

* Requests to localize a repository should point to a specific commit in the repository which needs to be localized.
    * The commit should contain XLF files which are up-to-date with respect to the original language files in the repository.
    * The commit should not contain any extraneous XLF files
    * The commit should not contain any extraneous translation units in the XLF files

## Workflow proposal

:memo: This section is written in terms of the [dotnet/roslyn](https://github.com/dotnet/roslyn) contributor workflow.

### General submissions

1. Pull requests target the **master** branch.
1. ~~The CI build automatically updates XLF files during the build.~~ (Unnecessary per https://github.com/dotnet/xliff-tasks/issues/42#issuecomment-350873963)
3. If the current build is a CI build for the **master** branch itself (as opposed to a build for a pull request), the build commits any updated XLF files and force-pushes those changes to the **localization/master** branch.
    * Pull requests do not result in changes to the XLF files in source control prior to merging.

#### Localization branch updates

Updates to XLF files occur during the build process when a special flag is set. These changes need to eventually be committed back to source control. The update and commit can occur in any of the following ways:

* Options for building
    * One of the existing Jenkins jobs can be run with the "update XLF" flag set
    * A new Jenkins job can be created to run the build with the flag set
    * The build can occur in a separate system which is responsible for this task
    * A developer (the localization tenet champ or a stand-in) builds the code manually
* Options for updating GitHub
    * An existing Jenkins job can be updated to push the changes back to GitHub
    * A new Jenkins job can be created to push the changes back to GitHub
    * An Azure function can be used which calls into the GitHub APIs to create new commit
    * A developer (the localization tenet champ or a stand-in) pushes the change to GitHub

My current preference for these operations would be the following:

1. A Jenkins job runs the build with the "update XLF" flag
2. The changes, if any, are attached to the build as an artifact
3. An Azure function uses the GitHub API to load the changes from the build artifact and apply them in a new commit on GitHub

### Localization process

1. A new branch **localize-master-*date*** is created at the current **localization/master** commit.
2. The newly-created branch is submitted for localization.
3. When the localization is complete, the changes are committed to **localize-master-*date*** and the branch is merged into **master**.

### Localization freeze

During a localization freeze of branch **master**, no changes are allowed which result in a change to localized resources. This needs to be enforced at the pull request level to avoid breaking the build. The pull request CI builds are updated to fail if the XLF update produces an output which does not match the existing XLF files in the **localization/master** branch. Due to the asynchronous nature of localizing files, it's important that the build *not* fail based solely on the presence of changes in XLF files during a build update.

**Comment by jaredpar** [comment]
> The CI build automatically updates XLF files during the build.

CI should never mutate sources. Ever.

**Comment by tannergooding** [comment]
The way XLF is designed currently is that CI should explicitly fail if a user forgets to modify sources (this helps ensure that we never merge a PR that would result in broken or incorrect state during our Signed build).

The XLF task should also be configured to automatically update the relevant XLF files when the repo is built locally.

So this should be a temporary issue while the changes to support XLF propagate through the branches.

That being said, it will also be an issue for users who rely entirely on Jenkins to validate changes and never build locally (but only users who are adding or updating resource files).

**Comment by sharwell** [comment]
> CI should never mutate sources. Ever.

The CI is expected to perform certain code generation tasks during the build, including but not limited to compilation tasks for XAML files. The changes proposed under this automation operate in the same manner. The location of the generated files is an unfortunate, but according to @nguerrera unavoidable, side effect of the way localization handoffs are performed. From a developer's perspective, the mutation required to be deterministic build-time code generation.

> The way XLF is designed currently is that CI should explicitly fail if a user forgets to modify sources

This fails the second developer goal.

> The XLF task should also be configured to automatically update the relevant XLF files when the repo is built locally.

This fails the first developer goal (specifically in regards to the *commit* and *review* items).

**Comment by tannergooding** [comment]
The difference between XAML code generation and the XLF generation is that the XLF files are required to be checked in for the LOC team to perform their work.

**Comment by jaredpar** [comment]
>  The changes proposed under this automation operate in the same manner. 

Disagree. All code generation today occurs into the output directory, not into the source tree.

**Comment by sharwell** [comment]
> Disagree. All code generation today occurs into the output directory, not into the source tree.

The goal can be equivalently achieved by having the build place updated XLF files in the intermediate output directory, and building against those instead of the original. I am neutral regarding the implementation on this aspect, but feel strongly that either way the generation process needs to be deterministic.

**Comment by tannergooding** [comment]
The XLF files are explicit source files, which cannot be dynamically generated, as they contain the actual localized text that is used when producing the satellite assemblies.

> Contributors should not be required to create, edit, commit, merge, or review XLF files unless they are contributing strings in a non-English language.

I believe this goal is a non-starter. Pull Requests need to be complete before being merged. This means tests and localization changes need to be included.

The localization changes should only ever be updating the metadata that tells the LOC tooling what needs to be updated.

If a user updates a resource file, it is reasonable to ask them to run a build command to update the corresponding metadata for the other languages so the LOC tooling can process it at a later date. I don't believe this is unreasonable.

**Comment by tannergooding** [comment]
> Requests to localize a repository should point to a specific commit in the repository which needs to be localized. 

This places a larger burden on both the LOC tenant owner and the LOC team itself. Rather than requesting they update a particular branch, we now have to track individual commits (which can be easily missed) and do that on a branch by branch basis.

**Comment by tannergooding** [comment]
> The goal can be equivalently achieved by having the build place updated XLF files in the intermediate output directory, and building against those instead of the original.

This won't work, as the XLF files contain localized strings that are checked in by the localization team.

**Comment by jaredpar** [comment]
> feel strongly that either way the generation process needs to be deterministic.

The current design is deterministic. 

> build commits any updated XLF files and force-pushes those changes to the localization/master branch.

This is a potential security problem. If CI has the capability to push to our repos then so does any PR. A devious user can simply find the credentials being used by CI and turn their PR into a push of whathever they choose.

**Comment by sharwell** [comment]
> The XLF files are explicit source files, which cannot be dynamically generated ...

The XLF files are already dynamically generated. The only question is whether the generation happens on the developer's machine or in an automated process that helps maintain the repository.

> This places a larger burden on both the LOC tenant owner and the LOC team itself.

This is generally acceptable. The entire automation design aims to avoid overhead where possible while emphasizing minimal impact of the localization process on external contributors. In the end, the localization tenant owner will be responsible for the remaining gaps. I believe that the gap itself can be quite small with a small amount of automation work in advance.

> This is a potential security problem. If CI has the capability to push to our repos then so does any PR.

I had some ideas regarding a hard separation between the CI and ability to push. Given how important this consideration is, I'll make sure to flesh these out.

> This won't work, as the XLF files contain localized strings that are checked in by the localization team.

I have seen this work in the past. Without getting too far into the details, the reason this works is the XLF needed by the build is a deterministic function of the original (English) language files and the XLF files currently in source control. The process simply needs to ensure that the builds used for a release have the deterministic function be an identity function of the XLF inputs.

> If a user updates a resource file, it is reasonable to ask them to run a build command to update the corresponding metadata for the other languages so the LOC tooling can process it at a later date. I don't believe this is unreasonable.

It is not fundamentally unreasonable, but as an additional requirement for external contributors (volunteers), it should be avoided if we have the means to do so. I believe we can create a solution with modest effort that achieves this, and have used a solution with very similar characteristics in the past.

**Comment by tannergooding** [comment]
> The XLF files are already dynamically generated. The only question is whether the generation happens on the developer's machine or in an automated process that helps maintain the repository.

I believe this is one source of confusion.

The XLF files are not "dynamically generated" (at least not in the classical sense). "Dynamic generation" (to me) implies that the file did not exist before and is either created or replaced in whole during build (as is the case with `*.designer.cs` files, and the like -- custom changes to these files are deleted when the file is next generated and will never be consumed by the build).

In the case of the XLF files they are never replaced in whole and are instead either created (if none exists) or partially updated (if it already exists -- custom changes to these files are preserved and will always be consumed by the build). Those files, however, need to exist both before and after and need to be checked into source control so that they can be consumed by tooling from the localization team to ensure that the culture specific texts are updated accordingly.

The default "state" of the world is that users would be required to explicitly update each non-English XLF file when adding a new or modifying an existing resource string. However, this is very automatable and we created tooling that will automatically perform these changes and make this easy on the developer.

**Comment by tmeschter** [comment]
Regarding this:

> Contributors should be able to submit updated translation(s) for a localization unit by modifying the relevant localization units in the XLF file(s) of interest, without incurring any other changes

That is not currently how a contributor would submit an updated translation. I'm not going to get into the details of the current process (as they aren't directly relevant to the larger points of this discussion) but the Loc team wishes to retain some oversight of the localized content without having to engage across a dozen or more different repos. We can discuss the process offline if you wish to know more, but you should remove this as a goal.

**Comment by tmeschter** [comment]
I would say the .xlf files are dynamically _updated_ rather than dynamically _generated_. They are a combination of data that _can_ be automatically generated by tooling (IDs and original English text) and data that _must_ be provided by a human (the translation).

**Comment by tmeschter** [comment]
Regarding the Localization Freeze:

>During a localization freeze of branch master, no changes are allowed which result in a change to localized resources. This needs to be enforced at the pull request level to avoid breaking the build. The pull request CI builds are updated to fail if the XLF update produces an output which does not match the existing XLF files in the localization/master branch. Due to the asynchronous nature of localizing files, it's important that the build not fail based solely on the presence of changes in XLF files during a build update.

As the World-Ready Tenet owner, I'll say that any process which involves regularly updating the CI process (especially on a per-branch basis, and especially _especially_ across a dozen-plus repos) to enforce policy is a non-starter. At best it's tedious busy work, and at worst we will forget to turn those policies on and something will get through.

So I'm going to say "no" to this approach.

**Comment by nguerrera** [comment]
> 2. The CI build automatically updates XLF files during the build.
> 3. If the current build is a CI build for the master branch itself (as opposed to a build for a pull request), the build commits any updated XLF files and force-pushes those changes to the localization/master branch

It is not necessary to update the XLF files during builds when there will be no commit of the result. The satellite build result of out-of-date XLF + up-to-date English source is the same as updating the XLF and then building. 

> The goal can be equivalently achieved by having the build place updated XLF files in the intermediate output directory, and building against those instead of the original

Again, this won't change the build result vs. just skipping the update step. 

The tooling in this repo is configured by default with `ErrorOnOutOfDateXlf=true`.  This is not to ensure that the build result is any different, but rather to ensure that the repo is always in a state that can be handed off for translation. Think of it this way: XLF is updated to influence the subsequent translation, not to influence the imminent build.

**Comment by nguerrera** [comment]
> This is a potential security problem

The updater can be a separate trusted process from CI, like the bot that merges branches. The one complication is that it needs to be able to run the `msbuild /t:UpdateXlf`. (Noting here that if the only goal a piece of automation is to update the XLF, it is not necessary to perform a full build only /t:UpdateXlf is needed). For Roslyn, I think this might not be especially hard since the msbuild dependency is already something that can be pulled down on to any machine, right?

**Comment by sharwell** [comment]
> As the World-Ready Tenet owner, I'll say that any process which involves regularly updating the CI process (especially on a per-branch basis, and especially especially across a dozen-plus repos) to enforce policy is a non-starter. At best it's tedious busy work, and at worst we will forget to turn those policies on and something will get through.

This is fine. The current implementation offers no automation to alert the tenet owner to a potential ship problem. The proposed automation change related to this had no purpose other than helping the tenet owner, so if you (as tenet owner) wouldn't find it helpful then the change can be omitted with no other side effects.

**Comment by sharwell** [comment]
> I would say the .xlf files are dynamically updated rather than dynamically generated. They are a combination of data that can be automatically generated by tooling (IDs and original English text) and data that must be provided by a human (the translation).

This is a technicality unrelated to the proposal. Regardless of how you describe the changes to these files, the basic requirement that edits to XLF files not be the responsibility of individual developers is the same. The only exception is cases where the developer is contributing a translation in a non-English language.

> It is not necessary to update the XLF files during builds when there will be no commit of the result.

I updated the original post to account for this.

> The updater can be a separate trusted process from CI, like the bot that merges branches. The one complication is that it needs to be able to run the msbuild /t:UpdateXlf. (Noting here that if the only goal a piece of automation is to update the XLF, it is not necessary to perform a full build only /t:UpdateXlf is needed). For Roslyn, I think this might not be especially hard since the msbuild dependency is already something that can be pulled down on to any machine, right?

It is also possible for the CI build to produce a patch which can then be applied by the bot. It *may* be possible to apply the patch using the GitHub API, without having to even clone the repository, but I'm not yet certain of this.

**Comment by sharwell** [comment]
> > The XLF files are already dynamically generated. ...
>
> I believe this is one source of confusion.

I was using the term 'generated' to refer to the fact that the user did not manually edit the file for the purpose of effecting a change on the resulting product. The change to the original source file (resx, vsct, etc.) was the only intended change.

I have also updated the wording of the original post to more accurately reflect the intended direction of this proposal. While a solution is not yet implemented or deployed, a resolution which adheres to the first two developer goals is a necessary condition of continuing to develop in collaboration with the open source community. This issue aims to facilitate discussions on additional goals and constraints the resulting automation operates under. In the limit, the fundamental goals (developer and workflow simplicity) of this proposal can be achieved by placing the XLF update responsibility on the world-ready tenet owner. However, I believe that limited automation can avoid many of these manual steps and have volunteered time to the implementation effort.

**Comment by CyrusNajmabadi** [comment]
Just to throw in my 2c, the new xlf system has imposed a significant burden on code reviews and the entire community process.  For example, here's a PR i'm trying to get through:

https://github.com/dotnet/roslyn/pull/21316/commits/3921ce341f668aaaf27791a1d75154da4167c1a6

It contains a few loc changes that end up expanding out to literally 1800 lines (or a full *80* pages) on github.  In other words, as someone reads the review, they will have to hit page-down/skip/whatever 80 times just to get through this stuff.  

This dramatically increases the noise, and decreases the signal that is so critical when people are doing PRs.

It seems wholly unfortunate, and extremely painful and inconvenient to force these artifacts to be part of the primary PR experience for users.

**Comment by CyrusNajmabadi** [comment]
An alternate solution would be the following: Do not require the dummy english strings to be placed into the foreign language xlf files.  Those strings are not needed as raw data that is checked into the repo.  Instead, they can be computed and generated for the files that delivered to the loc groups that actually do the localization. In other words, you can have:


1. English xlf file with all entries.
2. Foreign language xlf files, with only the entries that have been localized

From this, you can run a tool as part of a different process that takes the above and produces:

3. Foreign language xlf files, with the entries that have been localized, and the "needs localization" entries now added.

This batch of files can be loc'ed and hte merged back into the repo using whatever system we already have today for accepting loc changes back in.  

In other words, we do not need to check in data that indicates "the absence of data".  That absence can be automatically computed, by us, on demand.

**Comment by nguerrera** [comment]
That is no better than updating the xlf asynchronously as proposed, which I still support. As soon as you have a run-a-tool step then you lose the property that the repo is ready to hand off at any time. That can be fixed with automation, but as with everything, that has a cost that has to be balanced. Speaking as ex loc champ, it will not scale for such tooling to be run manually.


Also, the presence of the English source text in the xlf:

1. Is required by XLIFF standard and therefore expected by industry tools. We must provide valid XLIFF to our loc suppliers.

2. Allows detection of source text modification since last translation

3. Greatly facilitates community review of translations, which had been working out quite nicely.

**Comment by nguerrera** [comment]
Hmm, I think maybe you were just suggestimg the same asynchronous process I had in mind and they has been diacussed a lot; not excluding English text from xlf. Sorry if I misunderstood.

I've said this a bunch, but it seems to get lost in the specifics of various proposals: we don't have to update the XLF on every commit, but we must have the XLF updated before each hand-off. Doing it on every commit achieves this invariant without infrastructure to automate the updates asynchronously.

**Comment by nguerrera** [comment]
So just to be clear:

1. We don't need to write any new XLF updating logic or even a new tool. The logic in this repo is adequate, we would just want to run it at a more opportune time that doesn't mix in with code review PRs. xliff-tasks supports being run at any time, we just currently configure it to run on build and treat xliff that is not up-to-date as an error.

2. We don't need a new English XLF concept. The English text in source resx/vsct/xaml/unstructured files is is sufficent to inform the existing updater of how to update the translated XLF files.

The missing link is synchronizing updates with hand-offs.

**Comment by nguerrera** [comment]
> This batch of files can be loc'ed and hte merged back into the repo using whatever system we already have today for accepting loc changes back in.

The process for merging new translations is that the loc suppliers send a PR editing the XLF appropriately and we merge it.

The process for handing off a repo for localization is to point the loc suppliers at a repo URL + branch. There is a schedule with over a dozen repos/branches. Our end of the bargain is that these branch have all of the xlf files in the state they achieve from the updating logic we have now. 

The crux of the proposal here is the same as what I wrote [here](https://github.com/dotnet/roslyn/pull/23344#discussion_r154460289) under (3): "I'm thinking that we would commit the updates to a localization branch and then point the loc team at that branch."  Basically, we apply the xlf updates automatically and asynchronously to a separate branch, and hand off the separate branch.

**Comment by CyrusNajmabadi** [comment]
As another pain point: https://github.com/dotnet/roslyn/pull/21316

I now have conflicts in all the langauge xlf files that i have to go and manually fixup.  That's 13 files to fixup, even though i am not makign any relevant changes to those files (nor did the person i'm conflicting with).

**Comment by CyrusNajmabadi** [comment]
> That is no better than updating the xlf asynchronously as proposed, which I still support. As soon as you have a run-a-tool step then you lose the property that the repo is ready to hand off at any time. 

Why does it have to eb the *master* branch that is "ready to handoff at any time"?  Why can it not be some other branch?

In other words:

1. Devs make changes to master.  THey don't need to make xlf changes.
2. Tool runs after, detects that xlf updates are necessary.
3. Tool merges master into 'for-loc' branch, and runs whatever is necessary to spit out xlf files.
4. At some point in the future someone locs the changes in the 'for loc' branch and produces a PR.  That PR is merged in (either to master or 'for loc').

?

**Comment by CyrusNajmabadi** [comment]
> Speaking as ex loc champ, it will not scale for such tooling to be run manually.

I'm not asking for anything manual :).  This feels like a bog-standard github bot.  

But the point is that it's not part of the normal dev cycle to have to generate these pointless changes to files that aren't relevant to anything the dev is doing as part of their loop.  The entire update of foreign langauge xlf files can be done through another system/process that is independent of the dev loop.

**Comment by nguerrera** [comment]
> Why does it have to eb the master branch that is "ready to handoff at any time"? Why can it not be some other branch?

@CyrusNajmabadi You're basically +1ing what is already proposed here. See where I wrote "The crux of the proposal here is the same as what I wrote here under (3): "I'm thinking that we would commit the updates to a localization branch and then point the loc team at that branch." Basically, we apply the xlf updates automatically and asynchronously to a separate branch, and hand off the separate branch." See also that the proposal here documented by @sharwell involves updating a separate branch.

**Comment by nguerrera** [comment]
> I'm not asking for anything manual :). This feels like a bog-standard github bot.

Nobody is pushing back against an automated approach. Somebody has to take the time to implement it.

**Comment by CyrusNajmabadi** [comment]
Sorry, i guess i was confused by your responses.  I wrote something, and you stated:

> Speaking as ex loc champ, it will not scale for such tooling to be run manually.

It sounded like you were interpreting my posts as something that was requesting manual steps.  So i wanted to address that.  If you weren't actually referring to my posts, then my bad :)

**Comment by nguerrera** [comment]
No problem. FWIW, that was in reference to the sentence right before it:

> That can be fixed with automation, but as with everything, that has a cost that has to be balanced.

I was just saying that:

1.  There's development and maintenance costs to a bot-based solution. 
2.  The status quo is better than requiring any manual steps.

**Comment by CyrusNajmabadi** [comment]
I would agree with that.  Though the status-quo is pretty terribad right now.  Like a significant tax that i feel with nearly every PR that touches loc.  And i mean this for *any* PR, not just my PRs.  The entire experience around PRs has degraded significantly because of this current system.

**Comment by CyrusNajmabadi** [comment]
BTW, i have to deal with this every time i keep a branch up to date:

![image](https://user-images.githubusercontent.com/4564579/37503158-fbb60a34-2893-11e8-826f-e6a1396a8230.png)

It's just so obnoxiously terrible.  :(   Help me obi-wan-guerrara, you're my only hope.

Also, i would be willing to implement a roslyn feature for you in exchange for making this better.  Let me know if you're interested :)

**Comment by nguerrera** [comment]
Thanks for the "terribad" and "obnoxiously terrible" pejoratives. 😉 

The easiest way to fix such merge conflicts is to accept one side (ideally the side with more recent handback) and build again to re-update. 

I think it's worth taking a step back into the background of this project. When I took on the World Ready Tenet job, we had a bunch of issues in the black-box closed source localization process, and there was a new requirement to move to OSS and XLIFF. Among other reasons, we needed to build the full .NET Core stack including satellite assemblies from open source. 

I was pointed to some existing repos that had gone to XLIFF and they had several issues:

1. Tooling (mix of XliffParser and Multilingual App toolkit (MAT)) was Windows-only
2. Build integration was hard-coded into heterogenous build scripts in each repo.
3. (You'll love this) MAT actually edited 2X as many files as xliff-tasks: it synced en.resx to lang.xlf and back to lang.resx.
4. MAT had to be installed globally on the machine
5. Source file formats we needed like .vsct and CPS xaml rules were not supported.
6. MAT was not open-source

I set out to build something x-plat and OSS that you simply reference as a nupkg, and out come satellite assemblies and xlf files for the that loc team to translate. And that is exactly what the code in this repo achieves. 

Driving the XLF update process via bot+branch instead of the **optional** xliff-tasks UpdateXlfOnBuild=true we use now is a genuinely good idea. It is conceptually separate from this repo IMHO: it's building a service on top of these msbuild tasks. I did not have the bandwidth to go make that service and I especially did not have the bandwidth to maintain and monitor it.  The one nice thing about the current update approach is there's nothing behind the scenes to go down. CI forces a good state on every commit. If it breaks, it's in somebody's face and it gets fixed.

@sharwell had volunteered to drive this bot+branch approach and he has written is a good, detailed proposal here. If I were still the World Ready Tenet owner, I'd also want to see commitment from central engineering (@markwilkie and team) or @dotnet/roslyn-infrastructure for its ongoing maintenance. 

$0.02 CAD

**Comment by CyrusNajmabadi** [comment]
Another example (just to belabor the point as much as i can without pissing you off).  This PR has 95 pages of loc file changes: https://github.com/dotnet/roslyn/pull/23984

It's a full 50% of the review :eek: 😭

**Comment by nguerrera** [comment]
You don't need to convince me there's a problem worth solving. I have been supportive of this proposal since it's inception and I suggested the basic approach myself before Sam formalized it.

You are pissing me off.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by CyrusNajmabadi** [comment]
>It is conceptually separate from this repo IMHO: it's building a service on top of these msbuild tasks. 

I agree with the latter.  I'm not sure i agree with the former :)   Knowing how this xliff-task will be used, and how so many MS projects are impacted by moving to xliff (esp. if they're on github), it seems like a really good idea to sweat the details around the end user experience around this.

Right now, that experience is pretty bad *even* if it's light years ahead of where it was thanks to all those excellent pieces of work you mentioned above.  I don't think it's good to say: well, it's now in your hands, you can go write some scripts and make your life good again.  It think that this should be handled as part of the delivered product so that the teams that now have to use xlf can still have a good dev loop without every single one of them having to solve this problem for themselves.

**Comment by CyrusNajmabadi** [comment]
> You are pissing me off.

:)

**Comment by japj** [comment]
I have been looking into translation (build) workflow recently  and ‘discovered’ this repository (and MAT as well).
Great discussion!

One thing that I am wondering about, MAT has options to export/import xlf files.
I imagine this could be used to export only the strings needed for translation, send them to an agency and then when receiving translated content import it back into the “source” xlf.
Is that mechanism part of a translation workflow that is normally used? (It seems to be completely missing here)

Or are the xlf files send as a whole and then an agency will actually only update based on the state? (New/etc), and one just put the translated files as-is back into source control.

We currently have a lot of manual work figuring out which strings need to be updated, send them for translation and merge content back due to historical custom implementation of the translations.
It seems that xlf would be a good fit to solve some of the problems we face, but I’m also trying to understand the issues that you currently have and the route to tackle those.

**Comment by tmeschter** [comment]
@japj With XliffTasks the basic work flow is this:

1. Developers work entirely in .resx (or .vsct, or .txt, but from now on I'll just refer to .resx) files as they normally would, and generally do not touch .xlf files. The .resx are always the original source for translatable resources, and the job of the .xlf files is just to hold the translated equivalents.
2. XliffTasks automatically updates the .xlf files when the .resx changes--adding an item to the .resx adds an equivalent (with state="new") in the .xlf. Removing an item in the .resx removes the equivalent in the .resx. Updating an item that already has a translation (state="translated") in the .xlf resets the value of the `source` in the .xlf and sets the state to "needs-review-translation" to alert the translator to the change. The updating can be configured to occur on every build, or manually by running a specific MSBuild target.
3. The .xlf files sit alongside the other source files, are committed to version control (git, TFVC, etc.), and are versioned just like any other source file.
4. Translators periodically go through the .xlf files and add/update appropriate translations, and commit those changes.
5. XliffTasks modifies the build so that resource assemblies are generated from the .xlf files committed to the repo.

With this setup there's no manual effort to find the resources that need updates--you just check the .xlf files for elements with `state="new"` or `state="needs-review-translation"`.

In practice, many of the repos currently using XliffTasks (like [dotnet/roslyn](https://github.com/dotnet/roslyn)) are set to update the .xlf files on local build and devs are expected to commit the .xlf changes along with the original .resx changes. The CI build is set to fail if a .xlf file changes during its build (since that means a dev forgot something). This way a translator can come in at any time, pull the latest changes from the branch their working in, and be sure the .xlf files are up-to-date with the .resx files. There's no need for a separate export step. Similarly, there's no import step--the translator just creates a PR back into the branch with their changes.

**Comment by garath** [comment]
This repository has moved to "maintenance only" mode, though it remains important to .NET efforts. We're closing stale issues. Please feel free to reopen if desired and we'll triage the issue.
